### PR TITLE
Bump PHP to min version 7.2

### PR DESCRIPTION
**Pull request by phil-davis**
And associated bumps of phpunit etc. and refactoring of unit tests so they work with the new phpunit 8 and 9.

Fixes issue #176 (if this is agreed to go ahead)

https://phpunit.de/supported-versions.html

PHPUnit8 is for PHP 7.2+
PHPUnit9 is for PHP 7.3+

### CODE REVIEW

**Comment by phil-davis** [comment]
IMO this is good code.

But PR #174 could be reviewed and merged first - that moves the PHPUnit versions up a bit and gets CI running on PHP 7.4 using the existing PHP 5.6 minimum.

**Comment by ciaranmcnulty** [comment]
This looks good, can you add version 8 please?

**Comment by phil-davis** [comment]
@ciaranmcnulty done - please review.

IMO CI will pass. We are at the time-of-day when Travis starts getting a backlog of open-source builds: https://www.traviscistatus.com/#day

The "Backlog Linux Builds for Open Source projects" graph does not look bad yet. But Travis has not started the jobs yet.

**Comment by ciaranmcnulty** [comment]
I'll switch it to GitHub actions soon

**Comment by ciaranmcnulty** [comment]
Maybe you can rebase from master and apply your changes to the GitHub build instead?

**Comment by phil-davis** [comment]
CI just passed: https://travis-ci.org/github/Behat/Gherkin/builds/754479399

But now there is conflict because .travis.yml has been deleted.

I will try to rebase...

**Comment by phil-davis** [comment]
Done. The CI is fast! @ciaranmcnulty please review

**Comment by phil-davis** [comment]
Needs #184 sorted out first. Then I will rebase here...

**Comment by phil-davis** [comment]
@ciaranmcnulty I have rebased this.

I guess you will make a `4.7.0` release first - that will support PHP 5.6 through 7.4 and have the new feature #161 "Add support for multiple Example Tables with tags". That will allow people on old PHP versions to still use the new feature.

Then a `4.8.0` could come soon after, with this PR merged it will advertise "official" PHP 8.0 support.

**Comment by phil-davis** [comment]
@ciaranmcnulty is a release going to happen? And then merge this and release again?

**Comment by phil-davis** [comment]
Needs to wait until issue #187 PR #189 is sorted out and 4.7.1 is released. Then I will rebase this and someone can decide when they want to move this forward...

**Comment by phil-davis** [comment]
@ciaranmcnulty I have rebased and this is ready again, for whenever someone wants to do this.

But actually there is no rush. This PR adds PHP 8.0 to the test matrix in CI. But no actual changes to run-time code are needed. So we do know that the existing `master` run-time code does work with PHP 8.0. (It is just a bit harder if we want to routinely run the unit tests with all PHP from 5.6 through 8.0 in CI - the unit test code has to cope with a wide range of `phpunit` versions)

So maybe wait a while, in case other edge-cases or "stuff" needs to be fixed in the 4.7.* release series.

**Comment by ciaranmcnulty** [comment]
Yeah it’s tested with 8.0 via behat. As you say it’s more critical if we are doing maintenance 

I do have a PR coming based on this though because PHP 5 syntax annoys me so much

**Comment by phil-davis** [comment]
I personally have no problem with dropping the old PHP support. We are now always running our Behat test-runner with PHP 7.4, even if the back-end of the system-under-test happens to be running some other version of PHP (or some other programming language completely!).

I wonder how many people are actually still running their Behat test-runner scripts with "ancient" PHP?

**Comment by ciaranmcnulty** [comment]
Behat has dropped older versions but the major other consumer is Codeception and they support older PHP.

To be nice to them, we'd probably need to fix major bugs on 4.7.x for a while, it would be helpful if they aided in that :)

(ping @Naktibalda)

**Comment by Naktibalda** [comment]
I am afraid I've never written any Gherkin tests for Codeception.

We will release Codeception 5.0 in February which will require PHP 7.3, but Codeception 4 will receive some maintenance.

**Comment by stof** [comment]
@ciaranmcnulty if there is no BC break in gherkin, I'd rather release it as 4.8.

**Comment by ciaranmcnulty** [comment]
@Naktibalda Thanks, I don't feel so bad then! We can tag as 4.8 so Codeception 4 users will continue to use the older version and Codeception 5 get any newer features 👍

**Comment by phil-davis** [comment]
Rebased to fix a recent conflict in `composer.json`

**Comment by ciaranmcnulty** [comment]
So I guess we could merge this and call it `4.8.0`?

File: composer.json [hunk 1]
```diff
@@ -14,23 +14,23 @@
     ],
 
     "require": {
-        "php": ">=5.6"
+        "php": ">=7.2"
```
Comment (ciaranmcnulty) [inline composer.json:17]: I wonder if we should be more conservative here (i.e. `~7.2 | ~8.0`)

Comment (phil-davis) [reply composer.json:17]: I am happy to do that, because actually we have no idea if this code will work with PHP 9.0 !
I think someone told me not to do that in a previous review :)

Comment (phil-davis) [reply composer.json]: Done and rebased

**Comment by ciaranmcnulty** [comment]
Thanks @phil-davis

**Comment by phil-davis** [comment]
> So I guess we could merge this and call it `4.8.0`?

4.8.0 sounds good to me.
